src/fprime_extras/lint/rules/rule_valid_port_types.py

```diff
@@ -0,0 +1,162 @@
+            descr += " Please execute command from Fprime subdirectory or pass Fprime path as arg."
+            note = Notification(filename=extras_file.full_filename,
+                                rule_slug=slug,
+                                description=descr,
+                                line_number=0,
+                                column=0,
+                                severity=LintSeverity.WARNING
+                                )
+            notifications.append(note)
+        if doc_model.getroot().tag == 'component' and fprime_path is not None:
+            log.debug('This is a component file.')
+
+            top_tree = doc_model
+            top_root = top_tree.getroot()
+
+            component_namespace = top_root.get("namespace")
+
+            has_ports = False
+            # Make lists of tags
+            tag_lists_dict = {}
+            for child in top_root:
+                if child.tag not in tag_lists_dict:
+                    tag_lists_dict[child.tag] = []
+                    if child.tag == 'ports':
+                        has_ports = True
+                tag_lists_dict[child.tag].append(child)
+
+            if has_ports:
+                port_list = [p for p in tag_lists_dict['ports'][0] if p.tag == 'port']
+                import_list = tag_lists_dict['import_port_type']
+                imported_types, notes = self.get_imported_types(import_list, fprime_path, extras_file)
+                if len(notes) > 0:
+                    notifications.extend(notes)
+                    return notifications
+
+                for port in port_list:
+                    port_type = port.get('data_type')
+                    full_port_type = component_namespace + "::" + port_type
+                    # check if referenced type was imported. Check for local and global namespace
+                    valid = port_type in imported_types or full_port_type in imported_types
+                    if not valid:
+                        name = port.get('name')
+                        msg = 'Port "{}" specifies invalid type "{}". Imported types are {}.'
+                        msg = msg.format(name, port_type, imported_types)
+                        note = Notification(filename=extras_file.full_filename,
+                                            rule_slug=slug,
+                                            description=msg,
+                                            line_number=port.sourceline,
+                                            column=0,
+                                            severity=LintSeverity.CRITICAL
+                                            )
+                        notifications.append(note)
+        return notifications
+
+    def fix(self):
+        # if args.fix and flag:
+        #     conn.getparent().remove(conn)
+        pass
+
+    def reset(self):
+        pass
+
+    def get_imported_types(self, import_list, fprime_path, extras_file):
+        """
```

Comment:
Imported type errors should certainly be a different slug from the port kind slug.

Reply:
fixed this. There are now 3 slugs for this rule